src/genpy/generator.py

```diff
@@ -870,60 +878,56 @@ def msg_generator(msg_context, spec, search_path):
         yield ''
 
     if len(spec_names):
-        yield "  __slots__ = ['"+"','".join(spec_names)+"']"
-        yield "  _slot_types = ['"+"','".join(spec.types)+"']"
+        yield "  __slots__: List[str] = ['"+"','".join(spec_names)+"']"
+        yield "  _slot_types: List[str] = ['"+"','".join(spec.types)+"']"
     else:
-        yield '  __slots__ = []'
-        yield '  _slot_types = []'
+        yield '  __slots__: List[str] = []'
+        yield '  _slot_types: List[str] = []'
 
-    yield """
-  def __init__(self, *args, **kwds):
+    yield f"""
+  def __init__(self, {', '.join([f'{spec_name}: {format_spec_type_hint} = None' for spec_name, spec_type, format_spec_type_hint in fields])}):
     \"\"\"
     Constructor. Any message fields that are implicitly/explicitly
     set to None will be assigned a default value. The recommend
     use is keyword arguments as this is more robust to future message
     changes.  You cannot mix in-order arguments and keyword arguments.
 
     The available fields are:
-       %s
+      {','.join(spec_names)}
 
     :param args: complete set of field values, in .msg order
     :param kwds: use keyword arguments corresponding to message field names
     to set specific fields.
     \"\"\"
-    if args or kwds:
-      super(%s, self).__init__(*args, **kwds)""" % (','.join(spec_names), name)
-
+    super({name}, self).__init__(**{fields_dict})"""
+    # ^ TODO the above should use defualt values not = None. LEaving as is for the moment until everything is working
     if len(spec_names):
-        yield '      # message fields cannot be None, assign default values for those that are'
-        for (t, s) in zip(spec.types, spec_names):
-            yield '      if self.%s is None:' % s
-            yield '        self.%s = %s' % (s, default_value(msg_context, t, spec.package))
-    if len(spec_names) > 0:
-        yield '    else:'
-        for (t, s) in zip(spec.types, spec_names):
-            yield '      self.%s = %s' % (s, default_value(msg_context, t, spec.package))
+        yield '    # message fields cannot be None, assign default values for those that are'
 
+        for field in fields:
+            spec_name, spec_type, format_spec_type_hint = field
+            yield '    if self.%s is None:' % spec_name
+            yield INDENT*3 + 'self.%s: %s = %s' % (spec_name, format_spec_type_hint, default_value(msg_context, spec_type, spec.package))
     yield """
   def _get_types(self):
     \"\"\"
     internal API method
     \"\"\"
     return self._slot_types
 
-  def serialize(self, buff):
+  def serialize(self, buff: StringIO) -> None:
     \"\"\"
     serialize message into buffer
     :param buff: buffer, ``StringIO``
     \"\"\""""
     for y in serialize_fn_generator(msg_context, spec):
         yield '    ' + y
     yield """
-  def deserialize(self, str):
+  def deserialize(self, str: str) -> \'%s\':
```

Comment:
what does the return type `\'%s\' mean? Is it being replaced by something somehow?

Reply:
This is a old way of formatting python strings (before f-strings). 

It is worth noting that while functionally these two format methods are the same, the old template style string formatting is optimized in logging (and perhaps other libs) 

REF https://docs.python.org/3/howto/logging.html#optimization